drivers/flash/spi_nand.c

```diff
@@ -0,0 +1,1026 @@
+	if (is_addressed) {
+		union {
+			uint32_t u32;
+			uint8_t u8[4];
+		} addr32 = {
+			.u32 = sys_cpu_to_be32(addr),
+		};
+
+		if (access & NAND_ACCESS_32BIT_ADDR) {
+			address_len = 4;
+		} else if (access & NAND_ACCESS_24BIT_ADDR){
+			address_len = 3;
+		} else if (access & NAND_ACCESS_16BIT_ADDR){
+			address_len = 2;
+		} else if (access & NAND_ACCESS_8BIT_ADDR){
+			address_len = 1;
+		}
+		memcpy(&buf[1], &addr32.u8[3], address_len);
+		spi_buf[0].len += address_len;
+	};
+
+	if (access & NAND_ACCESS_DUMMY) {
+		spi_buf[0].len += 1;
+	}
+
+	const struct spi_buf_set tx_set = {
+		.buffers = spi_buf,
+		.count = (length != 0) ? 2 : 1,
+	};
+
+	const struct spi_buf_set rx_set = {
+		.buffers = spi_buf,
+		.count = 2,
+	};
+
+	if (is_write) {
+		return spi_write_dt(&driver_cfg->spi, &tx_set);
+	}
+
+	return spi_transceive_dt(&driver_cfg->spi, &tx_set, &rx_set);
+}
+
+#define spi_nand_cmd_read(dev, opcode, dest, length) \
+	spi_nand_access(dev, opcode, 0, 0, dest, length)
+#define spi_nand_cmd_write(dev, opcode) \
+	spi_nand_access(dev, opcode, NAND_ACCESS_WRITE, 0, NULL, 0)
+
+/* Everything necessary to acquire owning access to the device.
+ *
+ * This means taking the lock .
+ */
+static void acquire_device(const struct device *dev)
+{
+	if (IS_ENABLED(CONFIG_MULTITHREADING)) {
+		struct spi_nand_data *const driver_data = dev->data;
+
+		k_sem_take(&driver_data->sem, K_FOREVER);
+	}
+}
+
+/* Everything necessary to release access to the device.
+ *
+ * This means releasing the lock.
+ */
+static void release_device(const struct device *dev)
+{
+	if (IS_ENABLED(CONFIG_MULTITHREADING)) {
+		struct spi_nand_data *const driver_data = dev->data;
+
+		k_sem_give(&driver_data->sem);
+	}
+}
+
+
+static int spi_nand_get_feature(const struct device *dev,
+				uint8_t feature_addr, uint8_t *val)
+{
+        int ret = spi_nand_access(dev, SPI_NAND_CMD_GET_FEATURE,
+			NAND_ACCESS_ADDRESSED | NAND_ACCESS_8BIT_ADDR,
+			feature_addr, val, sizeof(*val));
+
+	return ret;
+}
+
+static int spi_nand_set_feature(const struct device *dev,
+				uint8_t feature_addr, uint8_t val)
+{
+        int ret = spi_nand_access(dev, SPI_NAND_CMD_SET_FEATURE,
+			NAND_ACCESS_WRITE | NAND_ACCESS_ADDRESSED | NAND_ACCESS_8BIT_ADDR,
+			feature_addr, &val, sizeof(val));
+
+	return ret;
+}
+
+/**
+ * @brief Wait until the flash is ready
+ *
+ * @note The device must be externally acquired before invoking this
+ * function.
+ *
+ * This function should be invoked after every ERASE, PROGRAM, or
+ * WRITE_STATUS operation before continuing.  This allows us to assume
+ * that the device is ready to accept new commands at any other point
+ * in the code.
+ *
+ * @param dev The device structure
+ * @return 0 on success, negative errno code otherwise
+ */
+static int spi_nand_wait_until_ready(const struct device *dev)
+{
+	int ret;
+	uint8_t reg = 0;
+
+	do {
+		ret = spi_nand_get_feature(dev, SPI_NAND_FEA_ADDR_STATUS, &reg);
+	} while (!ret && (reg & SPI_NAND_WIP_BIT));
```

Comment:
You might consider checking for other error conditions and timeout.  When I did this for a Kioxia part, the timeout was dependent on the operation.  Also, as I recall, the order of the checking is important as well.  That is, first check PROGF bit then ERASEF bit then OIP (WIP) then timeout.  I yielded during the wait because the wait can be quite a long time (block erase time can be on the order of several ms).